### PR TITLE
[MODULAR] Show DNR state for dead crew on crew monitor

### PR DESCRIPTION
**Pull request by tmyqlfpir**
## About The Pull Request

This PR adds a DNR state if a crew member is dead (and has suit sensors enabled).

## How This Contributes To The Skyrat Roleplay Experience

This is so medic personnel are able to make a choice to rescue a DNR body, or ignore them.

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary>
I have tested this on a local server and confirm that the DNR state only appears if the crew member in question is dead. These are all the suit sensor states with both dead and dnr shown.

Suit Sensors Living (1)
suit state 1 dead
suit state 1 dnr

Suit Sensors Vitals (2)
suit state 2 dead
suit state 2 dnr

Suit Sensors All (3)
suit state 3 dead
suit state 3 dnr
</details>

## Changelog

:cl:
fix: morgue incorrectly showing revivable state for dnr body
add: show dnr state to crew monitor for dead crew members
/:cl:

### CODE REVIEW

**Comment by Zergspower** [comment]
I made a completely modular SR version of the crewmonitor.js - please update that one instead which would also remove the incompliant tag.It's designed for both SR's healthpool and robotic players

**Comment by tmyqlfpir** [comment]
> I made a completely modular SR version of the crewmonitor.js - please update that one instead which would also remove the incompliant tag.It's designed for both SR's healthpool and robotic players

The crew monitor js this PR is adjusting is `CrewConsoleSkyrat.js`, what is the other js interface should I be changing instead?

**Comment by Zergspower** [comment]
weiiiird when i first looked at it - it was the other way around hah

**Comment by E231kei500bandai** [comment]
maybe it can have a different icon instead of being text?

**Comment by Zergspower** [comment]
Potentially yeah, @Zonespace27 would need to clarify what made it incompliant so you could make changes there too

File: code/game/machinery/computer/crew.dm [hunk 1]
```diff
@@ -261,6 +262,12 @@ GLOBAL_DATUM_INIT(crewmonitor, /datum/crewmonitor, new)
 			entry["is_robot"] = TRUE
 		// SKYRAT EDIT END
 
+		// SKYRAT EDIT BEGIN: Add DNR status
+		// If sensors are above living tracking, set DNR state
+		if (sensor_mode >= SENSOR_LIVING)
```
Comment (Zonespace27) [inline code/game/machinery/computer/crew.dm:264]: move this into the SR edit block on 260

**Review by Zonespace27 — CHANGES_REQUESTED** [review]
In the future please don't force-push commits, it nukes old history

**Comment by tmyqlfpir** [comment]
> In the future please don't force-push commits, it nukes old history

Thanks for the tip, I'll remember to do so for future PRs.

**Comment by tmyqlfpir** [comment]
The crew monitor will use the ban icon for unrevivable crew members as inspired by the tgstation PR linked above. Images of the crew monitor for every suit sensor state are shown in the first comment.

**Comment by Zergspower** [comment]
if i remember how i'll post how to fix the centering for the position, if not i'll be doing a  pass through that file in the next couple of weeks for an unrelated task to changing things and can make the update then

**Comment by Zonespace27** [comment]
https://github.com/tgstation/tgstation/pull/74006

**Comment by Gandalf2k15** [comment]
> [tgstation/tgstation#74006](https://github.com/tgstation/tgstation/pull/74006)

image

I actually agree with these changes, implement these

**Comment by tmyqlfpir** [comment]
> I actually agree with these changes, implement these

This is outside the scope of this PR, I simply wanted the crew monitor to display if a body can be recovered or not. If this is too OP it can be limited to positional tracking (max suit sensors) instead. Having an optional filter for the crew monitor would likely go unused, and would not solve the issue which is medic crew rescuing husked bodies.

And second, I do not know enough about DM language to implement such a feature, and will close instead. Thanks for reviewing this PR.

**Comment by E231kei500bandai** [comment]
Even if you made that "mark others unrevivable" feature, it wouldn't matter much anyway as you can already disable a body's suit sensors by just taking off their jumpsuit. Why make a new feature for something that can already be done?

**Comment by tmyqlfpir** [comment]
My main issue is seeing med staff repeatedly try to revive my DNR character because the morgue reports them as revivable, despite the DNR flag. It's an anti-frustration mechanic at best as I've seen other players spend 15+ minutes repairing my character's organs, only to have a DNR state when attempting to defib.

Also, it would be best if revivable state is only reported for tracking sensors given that someone being revivable is only useful if their location is known.

**Comment by E231kei500bandai** [comment]
I think you could probably make a new PR to add the morgue mechanic on its own. It's separate from the balance of the crew monitor.

**Comment by tmyqlfpir** [comment]
I agree - feel free to do that upstream. This has been overall frustrating for my first PR over a simple 14 LoC change.

**Comment by E231kei500bandai** [comment]
The DNR thing seems to already be a thing. If there's a ghost in the body, the light on the morgue slab is green. If there's no ghost in the body, the light is red. If you're DNR, you can't reenter your body so it has to be red. 
Perhaps it's not like that for bodies blacklisted at the end of a revolution, but that would be an exceptional scenario.

**Comment by tmyqlfpir** [comment]
>If you're DNR, you can't reenter your body so it has to be red.

It's a bug actually, if you have the DNR trait but don't set your body as DNR you can still reenter as the trait and state appear to be two different things. This extra check for morgue objects addresses it much like how if suicided will show red.

**Comment by E231kei500bandai** [comment]
How can you have the DNR trait and not be DNR? I think that's the bug that should befixed. 
By the way, TG has no DNR trait. So that would be a Skyrat issue.

**Comment by tmyqlfpir** [comment]
By setting your character with the DNR trait then dying, you will still be able to reenter your body. Examine text shows the character as still twitching despite DNR, and it doesn't really kick in unless you manually click DNR under ghost tab.

I know nothing of the internals of SS13 nor DM coding, but ideally the correct solution would be to check DNR trait on death, then apply DNR state to character automatically instead of requiring the player to manually click it afterwards so new medic players don't waste their time.

But given that the maintainers have not weighted in for over a month - this PR should be labeled DNR. I'm not going to follow up readdressing other issues after this, but feel free to.

**Comment by Zergspower** [comment]
> By setting your character with the DNR trait then dying, you will still be able to reenter your body. Examine text shows the character as still twitching despite DNR, and it doesn't really kick in unless you manually click DNR under ghost tab.
> 
> I know nothing of the internals of SS13 nor DM coding, but ideally the correct solution would be to check DNR trait on death, then apply DNR state to character automatically instead of requiring the player to manually click it afterwards so new medic players don't waste their time.
> 
> But given that the maintainers have not weighted in for over a month - this PR should be labeled DNR. I'm not going to follow up readdressing other issues after this, but feel free to.

Been trying to reach out to you but cant seem to do that - by chance is your account still open here?

**Comment by tmyqlfpir** [comment]
I am not interested in discussing this PR any further, requesting lock.

Edit: This PR has been eclipsed since by various PRs such as #23669, and is no longer relevant.

**Pull request closed without merging**